### PR TITLE
feature: improve `compose()` signature

### PR DESCRIPTION
**Pull request by layershifter**
This PR updates the compose signature to allow us to use it for creating as well as composing compoennts.

Examples:

```jsx
const Button = compose<'button', ButtonProps, ButtonStylesProps, {}, {}>(
  (props, ref, composeOptions) => {
    // component's code
  },
  {
    className: buttonClassName,
    displayName: 'Button',

    handledProps: [
      'accessibility',
      'as',
      'children',
      'circular',
      'className',
      'content',
      'styles',
      'variables',
    ],
  },
)

```

```jsx

const AttachmentAction = compose<
  'button',
  AttachmentActionOwnProps,
  AttachmentActionStylesProps,
  ButtonProps,
  ButtonStylesProps
>(Button, {
  className: attachmentActionClassName,
  displayName: 'AttachmentAction',
});
```

### CODE REVIEW

File: packages/react-compose/src/types.ts [hunk 1]
```diff
@@ -27,8 +27,13 @@ export type ComposedComponent<P = {}> = React.FunctionComponent<P> & {
   fluentComposeConfig: ComposePreparedOptions;
 };
 
+export type InputComposeComponent<P = {}> = React.FunctionComponent<P> & {
+  fluentComposeConfig?: ComposePreparedOptions;
+  defaultProps?: Partial<P>;
```
Comment (layershifter) [inline packages/react-compose/src/types.ts:32]:
```suggestion
```

This is defined in `FunctionComponent`

File: packages/react-compose/src/compose.ts [hunk 2]
```diff
@@ -1,67 +1,29 @@
...
-  Component.displayName = composeOptions.displayName || InputComponent.displayName;
-
-  Component.fluentComposeConfig = {
-    className: Component.deprecated_className,
-    displayNames: computeDisplayNames(inputOptions, composeOptions),
-
-    mapPropsToStylesPropsChain: (composeOptions.mapPropsToStylesProps
-      ? [...inputOptions.mapPropsToStylesPropsChain, composeOptions.mapPropsToStylesProps]
-      : inputOptions.mapPropsToStylesPropsChain) as ((props: ParentStylesProps & InputProps) => InputStylesProps)[],
+  if ((input as InputComposeComponent).defaultProps) {
```
Comment (mnajdova) [inline packages/react-compose/src/compose.ts:22]: @layershifter I guess this cast makes sense, as defaultProps will exists only on the components. Am I right?

File: packages/fluentui/react-northstar/src/components/Box/Box.tsx [hunk 3]
```diff
@@ -12,69 +20,76 @@ import {
...
-      styles,
-      variables,
-    }),
-    rtl: context.rtl,
-  });
+    const { accessibility, className, design, styles, variables, children, content } = props;
 
-  const unhandledProps = useUnhandledProps(Box.handledProps, props);
-  const ElementType = getElementType(props);
+    const getA11Props = useAccessibility(accessibility, {
```
Comment (pompomon) [inline packages/fluentui/react-northstar/src/components/Box/Box.tsx:47]:
```suggestion
    const getA11yProps = useAccessibility(accessibility, {
```
nit

File: packages/fluentui/react-northstar/src/components/Button/Button.tsx [hunk 4]
```diff
@@ -97,145 +99,191 @@ export type ButtonStylesProps = Pick<
...
-        getA11Props('loader', {
-          role: undefined,
-          styles: resolvedStyles.loader,
-        }),
+      design,
+    } = props;
+
+    const hasChildren = childrenExist(children);
+
+    const getA11Props = useAccessibility(accessibility, {
```
Comment (pompomon) [inline packages/fluentui/react-northstar/src/components/Button/Button.tsx:144]:
```suggestion
    const getA11yProps = useAccessibility(accessibility, {
```
nit

File: packages/react-compose/src/types.ts [hunk 5]
```diff
@@ -1,30 +1,64 @@
...
 
+//
+// Compose types
+//
+
+export type ComposedComponent<P = {}> = React.FunctionComponent<P> & {
+  fluentComposeConfig: ComposePreparedOptions;
+};
+
+export type InputComposeComponent<P = {}> = React.FunctionComponent<P> & {
```
Comment (pompomon) [inline packages/react-compose/src/types.ts:30]: What does this name mean? Is that "input-only" compose component, like form input elements only? Or the component which compose takes as an input?

Comment (mnajdova) [reply packages/react-compose/src/types.ts:30]: The component that compose takes as an input... I am open for some suggestions here to be honest ;)

File: packages/react-compose/src/compose.ts (hunk 2 above)
Comment (layershifter) [reply packages/react-compose/src/compose.ts:22]: Yep

File: packages/fluentui/react-northstar/src/components/Accordion/AccordionTitle.tsx [hunk 6]
```diff
@@ -137,37 +136,34 @@ class AccordionTitle extends UIComponent<WithAsProp<AccordionTitleProps>, any> {
   renderComponent({ ElementType, classes, unhandledProps, styles, accessibility }) {
     const { contentRef, children, content, indicator, contentWrapper } = this.props;
 
-    const contentWrapperElement = (
-      <Ref innerRef={contentRef}>
```
Comment (mnajdova) [inline packages/fluentui/react-northstar/src/components/Accordion/AccordionTitle.tsx:141]: This is suspicious, I had to remove the Ref and add the ref `contentRef` directly on the Box in order to solve the issue..

File: packages/fluentui/CHANGELOG.md [hunk 7]
```diff
@@ -53,6 +53,8 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Add `disabled` prop and styles to `Card` component @pompomon ([#12743](https://github.com/microsoft/fluentui/pull/12743))
 - Adding multiselection for the `Tree` @kolaps33, @assuncaocharles ([#12362](https://github.com/microsoft/fluentui/pull/12362))
 - Add `background` and `border` colors to `Card` in Teams theme @pompomon ([#12790](https://github.com/microsoft/fluentui/pull/12790))
+- Add compose options to `useStyles` @layershifter ([#12745](https://github.com/microsoft/fluentui/pull/12745))
+- Converted Box, Button* and Attachment* components to use compose @layershifter ([#12745](https://github.com/microsoft/fluentui/pull/12745))
```
Comment (layershifter) [inline packages/fluentui/CHANGELOG.md:57]:
```suggestion
- Add compose options to `useStyles` @layershifter @mnajdova ([#12745](https://github.com/microsoft/fluentui/pull/12745))
- Converted Box, Button* and Attachment* components to use compose @layershifter @mnajdova ([#12745](https://github.com/microsoft/fluentui/pull/12745))
```

Comment (mnajdova) [reply packages/fluentui/CHANGELOG.md:57]: :)

**Pull request merged**